lib/semantics/semantics.cc

```diff
@@ -228,6 +230,35 @@ void DoDumpSymbols(std::ostream &os, const Scope &scope, int indent) {
   --indent;
 }
 
+void Semantics::DumpSymbolsSources(std::ostream &os) {
+  NameToSymbolMap symbols;
+  GetAllSymbols(context_.globalScope(), symbols);
+  for (const auto pair : symbols) {
+    auto sourceInfo{cooked_.GetSourcePositionRange(pair.second->name())};
+    if (sourceInfo) {
+      os << pair.second->name().ToString() << ": " << sourceInfo->first.line
+         << ", " << sourceInfo->first.column << "-" << sourceInfo->second.column
+         << "\n";
+    } else if (pair.second->has<semantics::UseDetails>()) {
+      os << pair.second->name().ToString() << ": "
+         << pair.second->GetUltimate().owner().symbol()->name().ToString()
+         << "\n";
+    }
+  }
+}
+
+void GetAllSymbols(const Scope &scope, NameToSymbolMap &symbols) {
+  for (const auto &pair : scope) {
```

Comment:
You're taking all of the names in a scope and its children and flattening them into one `map` in such a way that conflicts are resolved in favor of `COMMON` blocks and the most deeply nested child scopes.

Reply:
I think this is fine because symbols with the same name resolve into the same source position and I am not using any other symbol information.

Reply:
If there is a variable `x` and a common block `/x/` in the same subprogram, clicking on a use of `x` will resolve to the common block, and that will always be wrong.  (19.3.2)

If there is a variable `x` in a subprogram and a variable `x` in one or more nested `BLOCK` constructs, whither will a click on a use of `x` resolve -- a declaration in the last `BLOCK`?

Reply:
Looking into the cases that you presented, in a subprogram with a variable `x` and a common block `/x/`, the function collects both the variable `x` and the common block `/x/` because they resolve into different locations. A similar thing happens for variable `x` in subprogram and variable `x` in `BLOCK` construct, both of them are collected into the map.

Example:
```
17 program main
18   integer :: x
19   integer :: y
20   common /x/ y
21   x = y
22 end program

f18  -fparse-only -fdebug-semantics  -fget-symbols-sources getsymbols04.f90
main: getsymbols04.f90, 17, 9-13
x: getsymbols04.f90, 18, 14-15
y: getsymbols04.f90, 19, 14-15
x: getsymbols04.f90, 20, 11-12
```
```
17 program main
18   integer :: x
19   integer :: y
20   block
21     integer :: x
22     x = y
23   end block
24 end program

f18  -fparse-only -fdebug-semantics  -fget-symbols-sources getsymbols05.f90
main: getsymbols05.f90, 17, 9-13
x: getsymbols05.f90, 18, 14-15
y: getsymbols05.f90, 19, 14-15
x: getsymbols05.f90, 21, 16-17
```

Reply:
Got it, you're comparing the values of `CharBlock`s rather than the contents of `CharBlock`s.